miscellaneous minor issues with implementation

Hello again! Thanks for your quick response to the last issue, and the new release :)

After the most recent update, I intended to implement ipware for an [actix-web](https://github.com/actix/actix-web) middleware, but I ran into a few other minor issues that unfortunately make the library more difficult to work with than I expected that I think would be fairly easy to resolve.

I'd be happy to make a PR covering all of these issues if that's acceptable. Would you prefer each one as a separate PR or if its fine if I update them all at once?

#### Issue 1:

The `dbg!` statements included in the code are showing up in my production web-server log:

```
[~/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/ipware-0.2.0/src/lib.rs:350:9] &self.proxy.proxy_list.len() = 0
```

According to rust [official documentation](https://doc.rust-lang.org/std/macro.dbg.html), `dbg!` is intended for short term debugging purposes only and should not ship into production. Their official statement is:

> Note that the macro is intended as a debugging tool and therefore you should avoid having uses of it in version control for long periods (other than in tests and similar). Debug output from production code is better done with other facilities such as the debug! macro from the log crate.

#### Issue 2:

The web-service crashes when given an invalid header due the [unwrap included in `get_client_ip`](https://github.com/orhanbalci/ipware/blob/main/src/lib.rs#L352). Preferably ipware should just ignore invalid headers and skip anything that's not valid utf-8.

I built a little actix-web POC to demonstrate this:

```rust
use actix_web::{App, HttpRequest, HttpServer, get};
use ipware::{IpWare, IpWareConfig, IpWareProxy};

#[get("/")]
async fn greet(req: HttpRequest) -> String {
    let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::default());
    let ip = ipware.get_client_ip(&req.headers().into(), true);
    format!("best ip {ip:?}")
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| App::new().service(greet))
        .bind(("localhost", 8000))?
        .run()
        .await
}
```

You can cause a panic by running the following curl command against the server: 

`curl 'http://localhost:8000/' -H 'X-Forwarded-For: 😃'`

#### Issue 3:

`IpWareProxy/IpWare` structs are not portable due to their associated lifetime.

In my specific instance, I'm trying to implement [middleware for actix-web](https://actix.rs/docs/middleware) which requires a struct to implement specific traits. Due to the constraints of this trait I cannot have a lifetime tied to it or any of its associated fields, meaning I'm forced to re-construct `IpWareProxy/IpWare` for every instance that I call it, making the code less performant and more verbose.

In my case, and I think arguably all cases that I can think of, users of ipware will most likely want to pre-construct `IpWare` or at minimum `IpWareProxy` ahead of time and re-use the same object since its unlikely for ipware's configuration to need to be dynamic per-request, especially since the lifetime requirement is on `IpWareProxy` which is supposed to be a specification on trust for an upstream proxy, which is unlikely to change.

Removing the lifetime and using a `Vec<IpAddr>` over `&[IpAddr]` not only makes object construction and re-use simpler, but also allows you to add additional features like a `Default` impl for `IpWare` which is currently missing.

> Anyways you can update if it will be easier for you. Single PR is fine.

Alright, I'll upload soon thank you :)

> On the third issue can you state which traits should be implemented and why lifetimes are an issue. I'm not familiar with actix-web.

Actix-web has a lot of documentation you can view on middleware implementation available on their website: https://actix.rs/docs/middleware/

An example implementation looks something like this, where the middleware is first constructed with a factory implementing the [`Transform`](https://docs.rs/actix-web/latest/actix_web/dev/trait.Transform.html), that then transforms into a service object that implements the [`Service`](https://docs.rs/actix-web/latest/actix_web/dev/trait.Service.html) trait.

<details>
  <summary>Middleware Example</summary>

```rust
use std::future::{ready, Ready};

use actix_web::{
    dev::{forward_ready, Service, ServiceRequest, ServiceResponse, Transform},
    Error,
};
use futures_util::future::LocalBoxFuture;

// There are two steps in middleware processing.
// 1. Middleware initialization, middleware factory gets called with
//    next service in chain as parameter.
// 2. Middleware's call method gets called with normal request.
pub struct SayHi;

// Middleware factory is `Transform` trait
// `S` - type of the next service
// `B` - type of response's body
impl<S, B> Transform<S, ServiceRequest> for SayHi
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type InitError = ();
    type Transform = SayHiMiddleware<S>;
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ready(Ok(SayHiMiddleware { service }))
    }
}

pub struct SayHiMiddleware<S> {
    service: S,
}

impl<S, B> Service<ServiceRequest> for SayHiMiddleware<S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;

    forward_ready!(service);

    fn call(&self, req: ServiceRequest) -> Self::Future {
        println!("Hi from start. You requested: {}", req.path());

        let fut = self.service.call(req);

        Box::pin(async move {
            let res = fut.await?;

            println!("Hi from response");
            Ok(res)
        })
    }
}
```

</details>

You can see in both separate traits, that they rely on a `'static` lifetime. If I try and construct the `IpWare` struct during the service transformation, I'm forced to assign a lifetime to the `SayHiMiddleware` for example.

<details>
  <summary>IpWare Construction during Transform</summary>

```rust
impl<'a, S, B> Transform<S, ServiceRequest> for SayHi
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type InitError = ();
    type Transform = SayHiMiddleware<'a, S>;
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ready(Ok(SayHiMiddleware {
            service,
            ipware: IpWare::new(
                IpWareConfig::default(),
                IpWareProxy::new(0, &[IpAddr::from_str("1.2.3.4").unwrap()]),
            ),
        }))
    }
}

pub struct SayHiMiddleware<'a, S> {
    service: S,
    ipware: IpWare<'a>,
}
```

</details>

However, this causes the following error:

```
error[E0207]: the lifetime parameter `'a` is not constrained by the impl trait, self type, or predicates
  --> src/main.rs:21:6
   |
21 | impl<'a, S, B> Transform<S, ServiceRequest> for SayHi
   |      ^^ unconstrained lifetime parameter
```

If I set the lifetime to static I also get an error:

```
error[E0515]: cannot return value referencing temporary value
  --> src/main.rs:38:9
   |
38 | /         ready(Ok(SayHiMiddleware {
39 | |             service,
40 | |             ipware: IpWare::new(
41 | |                 IpWareConfig::default(),
42 | |                 IpWareProxy::new(0, &[IpAddr::from_str("1.2.3.4").unwrap()]),
   | |                                      -------------------------------------- temporary value created here
43 | |             ),
44 | |         }))
   | |___________^ returns a value referencing data owned by the current function
```

Even if we implement `Clone` on `IpWareProxy` to allow it to be passed during construction and add lifetimes all the way through the scope of the middleware in the example below I get the following errors:

<details>
  <summary>Complete Middleware with Lifetimes</summary>


```rust
use actix_web::{App, HttpRequest, HttpServer, get};
use ipware::{IpWare, IpWareConfig, IpWareProxy};

use std::{
    future::{Ready, ready},
    net::IpAddr,
    str::FromStr,
};

use actix_web::{
    Error,
    dev::{Service, ServiceRequest, ServiceResponse, Transform, forward_ready},
};
use futures_util::future::LocalBoxFuture;

// There are two steps in middleware processing.
// 1. Middleware initialization, middleware factory gets called with
//    next service in chain as parameter.
// 2. Middleware's call method gets called with normal request.
pub struct SayHi<'a> {
    proxy: IpWareProxy<'a>,
}

// Middleware factory is `Transform` trait
// `S` - type of the next service
// `B` - type of response's body
impl<'a, S, B> Transform<S, ServiceRequest> for SayHi<'a>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type InitError = ();
    type Transform = SayHiMiddleware<'a, S>;
    type Future = Ready<Result<Self::Transform, Self::InitError>>;

    fn new_transform(&self, service: S) -> Self::Future {
        ready(Ok(SayHiMiddleware {
            service,
            ipware: IpWare::new(IpWareConfig::default(), self.proxy.clone()),
        }))
    }
}

pub struct SayHiMiddleware<'a, S> {
    service: S,
    ipware: IpWare<'a>,
}

impl<'a, S, B> Service<ServiceRequest> for SayHiMiddleware<'a, S>
where
    S: Service<ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
    S::Future: 'static,
    B: 'static,
{
    type Response = ServiceResponse<B>;
    type Error = Error;
    type Future = LocalBoxFuture<'static, Result<Self::Response, Self::Error>>;

    forward_ready!(service);

    fn call(&self, req: ServiceRequest) -> Self::Future {
        println!("Hi from start. You requested: {}", req.path());

        let fut = self.service.call(req);

        Box::pin(async move {
            let res = fut.await?;

            println!("Hi from response");
            Ok(res)
        })
    }
}

#[get("/")]
async fn greet(req: HttpRequest) -> String {
    let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::default());
    let ip = ipware.get_client_ip(&req.headers().into(), true);
    format!("best ip {ip:?}")
}

#[actix_web::main]
async fn main() -> std::io::Result<()> {
    let proxy = IpWareProxy::new(0, &[IpAddr::from_str("1.2.3.4").unwrap()]);

    HttpServer::new(|| App::new().service(greet).wrap(SayHi { proxy }))
        .bind(("localhost", 8000))?
        .run()
        .await
}
```

</details>

```
error[E0507]: cannot move out of `proxy`, a captured variable in an `Fn` closure
  --> src/main.rs:89:63
   |
87 |     let proxy = IpWareProxy::new(0, &[IpAddr::from_str("1.2.3.4").unwrap()]);
   |         ----- captured outer variable
88 |
89 |     HttpServer::new(|| App::new().service(greet).wrap(SayHi { proxy }))
   |                     -- captured by this `Fn` closure          ^^^^^ move occurs because `proxy` has type `IpWareProxy<'_>`, which does not implement the `Copy` trait

error[E0716]: temporary value dropped while borrowed
  --> src/main.rs:87:38
   |
87 |     let proxy = IpWareProxy::new(0, &[IpAddr::from_str("1.2.3.4").unwrap()]);
   |                                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ - temporary value is freed at the end of this statement
   |                                      |
   |                                      creates a temporary value which is freed while still in use
88 |
89 |     HttpServer::new(|| App::new().service(greet).wrap(SayHi { proxy }))
   |                        ----------------------------------------------- argument requires that borrow lasts for `'static`
```

--- a/README.md
+++ b/README.md
@@ -103,7 +103,7 @@ let request_headers_precedence = vec![
 ```
 
 You can customize the order by providing your own list using IpWareConfig.
-```rust no_run
+```rust
 use ipware::IpWareConfig;
 use http::HeaderName;
 // specific header name
@@ -111,12 +111,12 @@ IpWareConfig::new(vec![HeaderName::from_static("http_x_forwarded_for")],true);
 
 // multiple header names
 IpWareConfig::new(
-               vec![
-                   HeaderName::from_static("http_x_forwarded_for"),
-                   HeaderName::from_static("x_forwarded_for"),
-               ],
-               true,
-           );
+    vec![
+        HeaderName::from_static("http_x_forwarded_for"),
+        HeaderName::from_static("x_forwarded_for"),
+    ],
+    true,
+);
 ```
 
 #### 🤝 Trusted Proxies
@@ -127,20 +127,18 @@ by providing a `trusted proxy list`, or a known proxy `count`.
 You can customize the proxy IP prefixes by providing your own list by using IpWareProxy.
 You can pass your custom list on every call, when calling the proxy-aware api to fetch the ip.
 
-```rust no_run
+```rust
 // In the above scenario, use your load balancer IP address as a way to filter out unwanted requests.
 use std::net::IpAddr;
-use ipware::IpWare;
-use ipware::IpWareConfig;
-use ipware::IpWareProxy;
-use http::HeaderMap;
+
+use ipware::{HeaderMap, IpWare, IpWareConfig, IpWareProxy};
 
 let headers = HeaderMap::new(); // replace this with your own headers
 let proxies = vec![
-            "198.84.193.157".parse::<IpAddr>().unwrap(),
-            "198.84.193.158".parse::<IpAddr>().unwrap(),
-        ];
-let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+    "198.84.193.157".parse::<IpAddr>().unwrap(),
+    "198.84.193.158".parse::<IpAddr>().unwrap(),
+];
+let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
 
 // usage: non-strict mode (X-Forwarded-For: <fake>, <client>, <proxy1>, <proxy2>)
 // The request went through our <proxy1> and <proxy2>, then our server
@@ -159,7 +157,7 @@ let (ip, trusted_route) = ipware.get_client_ip(&headers, true);
 If your http server is behind a `known` number of proxies, but you deploy on multiple providers and don't want to track proxy IPs, you still can filter out unwanted requests by providing proxy `count`.
 
 You can customize the proxy count by providing your `proxy_count` using IpWareProxy.
-```rust no_run
+```rust
 use ipware::*;
 use std::net::IpAddr;
 
@@ -168,11 +166,11 @@ use std::net::IpAddr;
 
 let headers = HeaderMap::new(); // replace this with your own headers
 let proxies = vec![];
-let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
 
 // enforce proxy count and trusted proxies
 let proxies = vec!["198.84.193.157".parse::<IpAddr>().unwrap()];
-let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
 
 // usage: non-strict mode (X-Forwarded-For: <fake>, <client>, <proxy1>, <proxy2>)
 // total number of ip addresses are greater than the total count
@@ -197,7 +195,7 @@ for the originating client IP address is the `leftmost`as per`client, proxy1, pr
 trusted proxy.
 However, in rare cases your network has a `custom` configuration where the `rightmost` IP address is that of the originating client. If that is the case, then indicate it when creating:
 ```
-```rust no_run
+```rust
 use ipware::*;
 let ipware = IpWare::new(
     IpWareConfig::default().leftmost(false),

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -347,23 +347,19 @@ impl<'a> IpWare<'a> {
         let mut loopback_list = vec![];
         let mut private_list = vec![];
         let meta_values = self.get_meta_values(headers);
-        dbg!(&self.proxy.proxy_list.len());
         for &meta_value in meta_values.iter() {
             let meta_ips = self.get_ips_from_string(meta_value.to_str().unwrap().to_owned());
             if meta_ips.is_empty() {
                 continue;
             }
             let proxy_count_validated = self.proxy.is_proxy_count_valid(&meta_ips, strict);
-            dbg!(&proxy_count_validated);
             if !proxy_count_validated {
                 continue;
             }
             let proxy_list_validated = self.proxy.is_proxy_trusted_list_valid(&meta_ips, strict);
-            dbg!(&proxy_list_validated);
             if !proxy_list_validated {
                 continue;
             }
-            dbg!(meta_ips.clone());
             let (client_ip, trusted_route) =
                 self.get_best_ip(&meta_ips, proxy_count_validated, proxy_list_validated);
             if let Some(client_ip) = client_ip {
@@ -429,7 +425,6 @@ impl<'a> IpWare<'a> {
         if ip_list.is_empty() {
             return (None, false);
         }
-        dbg!(&self.proxy.proxy_list);
         if !self.proxy.proxy_list.is_empty() && proxy_list_validated {
             return (ip_list.iter().rev().nth(self.proxy.proxy_list.len()), true);
         }

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -127,10 +127,8 @@
 //! ```rust
 //! // In the above scenario, use your load balancer IP address as a way to filter out unwanted requests.
 //! use std::net::IpAddr;
-//! use ipware::IpWare;
-//! use ipware::IpWareConfig;
-//! use ipware::IpWareProxy;
-//! use http::HeaderMap;
+//!
+//! use ipware::{HeaderMap, IpWare, IpWareConfig, IpWareProxy};
 //!
 //! let headers = HeaderMap::new(); // replace this with your own headers
 //! let proxies = vec![
@@ -149,7 +147,7 @@
 //! // Total ip address are total trusted proxies + client ip
 //! // We don't allow far-end proxies, or fake addresses (exact or None)
 //! let (ip, trusted_route) = ipware.get_client_ip(&headers, true);
-//! ```
+//!  ```
 //!
 //! ### Proxy Count
 //!
@@ -157,8 +155,8 @@
 //!
 //! You can customize the proxy count by providing your `proxy_count` using IpWareProxy.
 //! ```rust
-//! use ipware::*;
 //! use std::net::IpAddr;
+//! use ipware::{HeaderMap, IpWare, IpWareConfig, IpWareProxy};
 //!
 //! // In the above scenario, the total number of proxies can be used as a way to filter out unwanted requests.
 //! // enforce proxy count
@@ -195,7 +193,8 @@
 //! However, in rare cases your network has a `custom` configuration where the `rightmost` IP address is that of the originating client. If that is the case, then indicate it when creating:
 //! ```
 //!```rust
-//! use ipware::*;
+//! use ipware::{IpWare, IpWareConfig, IpWareProxy};
+//!
 //! let ipware = IpWare::new(
 //!     IpWareConfig::default().leftmost(false),
 //!     IpWareProxy::default(),
@@ -203,10 +202,10 @@
 //! ```
 
 use std::net::{IpAddr, SocketAddr};
+use std::str::FromStr;
 use std::string::ToString;
 
-pub use http::header::*;
-pub use http::HeaderName;
+pub use http::header::{HeaderMap, HeaderName, HeaderValue};
 
 #[derive(Clone, Debug)]
 pub struct IpWareConfig {
@@ -264,53 +263,52 @@ pub struct IpWareProxy {
 }
 
 impl IpWareProxy {
-    pub fn new(proxy_count: u16, proxy_list: Vec<IpAddr>) -> Self {
-        IpWareProxy { proxy_count, proxy_list }
+    pub fn new<T>(proxy_count: u16, proxy_list: T) -> Self
+    where
+        T: Into<Vec<IpAddr>>,
+    {
+        IpWareProxy { proxy_count, proxy_list: proxy_list.into() }
     }
 
-    pub fn is_proxy_count_valid<'a, I: IntoIterator<Item = &'a IpAddr>>(
-        &self,
-        ip_list: I,
-        strict: bool,
-    ) -> bool {
+    pub fn is_proxy_count_valid<'a, I>(&self, ip_list: I, strict: bool) -> bool
+    where
+        I: IntoIterator<Item = &'a IpAddr>,
+    {
         if self.proxy_count < 1 {
-            true
-        } else {
-            let ip_count = ip_list.into_iter().collect::<Vec<_>>().len();
-
-            if ip_count < 1 {
-                false
-            } else if strict {
-                self.proxy_list.len() == ip_count - 1
-            } else {
-                ip_count - 1 > self.proxy_list.len()
-            }
+            return true;
         }
+
+        let ip_count = ip_list.into_iter().collect::<Vec<_>>().len();
+        if ip_count < 1 {
+            return false;
+        }
+        if strict {
+            return self.proxy_list.len() == ip_count - 1;
+        }
+
+        ip_count - 1 > self.proxy_list.len()
     }
 
-    pub fn is_proxy_trusted_list_valid<'a, I: IntoIterator<Item = &'a IpAddr>>(
-        &self,
-        ip_list: I,
-        strict: bool,
-    ) -> bool {
+    pub fn is_proxy_trusted_list_valid<'a, I>(&self, ip_list: I, strict: bool) -> bool
+    where
+        I: IntoIterator<Item = &'a IpAddr>,
+    {
         if self.proxy_list.is_empty() {
-            true
-        } else {
-            let ip_list = ip_list.into_iter().collect::<Vec<_>>();
-            let ip_count = ip_list.len();
-            let proxy_count = self.proxy_list.len();
-            if (strict && ip_count - 1 != proxy_count) || (ip_count - 1 < proxy_count) {
-                false
-            } else {
-                ip_list
-                    .into_iter()
-                    .rev()
-                    .take(proxy_count)
-                    .rev()
-                    .zip(self.proxy_list.iter())
-                    .all(|(ip_addr, proxy_addr)| *ip_addr == *proxy_addr)
-            }
+            return true;
         }
+        let ip_list = ip_list.into_iter().collect::<Vec<_>>();
+        let ip_count = ip_list.len();
+        let proxy_count = self.proxy_list.len();
+        if (strict && ip_count - 1 != proxy_count) || (ip_count - 1 < proxy_count) {
+            return false;
+        }
+        ip_list
+            .into_iter()
+            .rev()
+            .take(proxy_count)
+            .rev()
+            .zip(self.proxy_list.iter())
+            .all(|(ip_addr, proxy_addr)| ip_addr == proxy_addr)
     }
 }
 
@@ -330,9 +328,8 @@ impl IpWare {
         headers: &'a HeaderMap,
         name: &HeaderName,
     ) -> Option<&'a HeaderValue> {
-        let value = headers.get(name);
-        match value {
-            Some(_) => value,
+        match headers.get(name) {
+            Some(value) => Some(value),
             None => headers.get(name.to_string().replace('_', "-")),
         }
     }
@@ -343,9 +340,10 @@ impl IpWare {
             .iter()
             .filter_map(|header_name| self.get_meta_value(headers, header_name))
             .filter_map(|header_value| header_value.to_str().ok())
-            .collect::<Vec<_>>()
+            .collect()
     }
 
+    /// Returns the client's IP address.
     pub fn get_client_ip(&self, headers: &HeaderMap, strict: bool) -> (Option<IpAddr>, bool) {
         let mut loopback_list = vec![];
         let mut private_list = vec![];
@@ -394,29 +392,21 @@ impl IpWare {
     /// # Arguments
     /// * `ip_str` - String contains , seperated ip addresses
     fn get_ips_from_string(&self, ip_str: &str) -> Vec<IpAddr> {
-        let mut result = ip_str
+        let Ok(mut result): Result<Vec<_>, _> = ip_str
             .split(',')
-            .map(|single_ip| {
-                let trimmed_ip = single_ip.trim_start().trim_end();
-                let maybe_ipaddr = trimmed_ip.parse::<IpAddr>();
-                match maybe_ipaddr {
-                    Ok(_) => maybe_ipaddr.ok(),
-                    Err(_) => trimmed_ip
-                        .parse::<SocketAddr>()
-                        .map(|socket_addr| socket_addr.ip())
-                        .ok(),
-                }
+            .map(|single_ip| single_ip.trim())
+            .map(|trimmed_ip| match IpAddr::from_str(trimmed_ip) {
+                Ok(ip) => Ok(ip),
+                Err(_) => SocketAddr::from_str(trimmed_ip).map(|socket_addr| socket_addr.ip()),
             })
-            .collect::<Vec<_>>();
-        if result.iter().any(|maybe_ipaddr| maybe_ipaddr.is_none()) {
-            return vec![];
-        } else if !self.config.leftmost {
+            .collect()
+        else {
+            return Vec::new();
+        };
+        if !self.config.leftmost {
             result.reverse();
         }
         result
-            .iter()
-            .map(|maybe_ipaddr| maybe_ipaddr.unwrap())
-            .collect::<Vec<_>>()
     }
 
     fn get_best_ip<'a>(
@@ -438,7 +428,7 @@ impl IpWare {
                 true,
             );
         }
-        return (ip_list.first(), false);
+        (ip_list.first(), false)
     }
 }
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -137,7 +137,7 @@
 //!             "198.84.193.157".parse::<IpAddr>().unwrap(),
 //!             "198.84.193.158".parse::<IpAddr>().unwrap(),
 //!         ];
-//! let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+//! let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
 //!
 //! // usage: non-strict mode (X-Forwarded-For: <fake>, <client>, <proxy1>, <proxy2>)
 //! // The request went through our <proxy1> and <proxy2>, then our server
@@ -165,11 +165,11 @@
 //!
 //! let headers = HeaderMap::new(); // replace this with your own headers
 //! let proxies = vec![];
-//! let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+//! let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
 //!
 //! // enforce proxy count and trusted proxies
 //! let proxies = vec!["198.84.193.157".parse::<IpAddr>().unwrap()];
-//! let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+//! let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
 //!
 //! // usage: non-strict mode (X-Forwarded-For: <fake>, <client>, <proxy1>, <proxy2>)
 //! // total number of ip addresses are greater than the total count
@@ -208,6 +208,7 @@ use std::string::ToString;
 pub use http::header::*;
 pub use http::HeaderName;
 
+#[derive(Clone, Debug)]
 pub struct IpWareConfig {
     precedence: Vec<HeaderName>,
     leftmost: bool,
@@ -256,18 +257,18 @@ impl IpWareConfig {
     }
 }
 
-#[derive(Default)]
-pub struct IpWareProxy<'a> {
+#[derive(Clone, Debug, Default)]
+pub struct IpWareProxy {
     proxy_count: u16,
-    proxy_list: &'a [IpAddr],
+    proxy_list: Vec<IpAddr>,
 }
 
-impl<'a> IpWareProxy<'a> {
-    pub fn new(proxy_count: u16, proxy_list: &'a [IpAddr]) -> Self {
+impl IpWareProxy {
+    pub fn new(proxy_count: u16, proxy_list: Vec<IpAddr>) -> Self {
         IpWareProxy { proxy_count, proxy_list }
     }
 
-    pub fn is_proxy_count_valid<'b, I: IntoIterator<Item = &'b IpAddr>>(
+    pub fn is_proxy_count_valid<'a, I: IntoIterator<Item = &'a IpAddr>>(
         &self,
         ip_list: I,
         strict: bool,
@@ -287,7 +288,7 @@ impl<'a> IpWareProxy<'a> {
         }
     }
 
-    pub fn is_proxy_trusted_list_valid<'b, I: IntoIterator<Item = &'b IpAddr>>(
+    pub fn is_proxy_trusted_list_valid<'a, I: IntoIterator<Item = &'a IpAddr>>(
         &self,
         ip_list: I,
         strict: bool,
@@ -313,21 +314,22 @@ impl<'a> IpWareProxy<'a> {
     }
 }
 
-pub struct IpWare<'a> {
+#[derive(Clone, Debug, Default)]
+pub struct IpWare {
     config: IpWareConfig,
-    proxy: IpWareProxy<'a>,
+    proxy: IpWareProxy,
 }
 
-impl<'a> IpWare<'a> {
-    pub fn new(config: IpWareConfig, proxy: IpWareProxy<'a>) -> Self {
+impl IpWare {
+    pub fn new(config: IpWareConfig, proxy: IpWareProxy) -> Self {
         IpWare { config, proxy }
     }
 
-    fn get_meta_value<'b>(
+    fn get_meta_value<'a>(
         &self,
-        headers: &'b HeaderMap,
+        headers: &'a HeaderMap,
         name: &HeaderName,
-    ) -> Option<&'b HeaderValue> {
+    ) -> Option<&'a HeaderValue> {
         let value = headers.get(name);
         match value {
             Some(_) => value,
@@ -335,7 +337,7 @@ impl<'a> IpWare<'a> {
         }
     }
 
-    fn get_meta_values(&self, headers: &'a HeaderMap) -> Vec<&'a str> {
+    fn get_meta_values<'a>(&self, headers: &'a HeaderMap) -> Vec<&'a str> {
         self.config
             .precedence
             .iter()
@@ -417,12 +419,12 @@ impl<'a> IpWare<'a> {
             .collect::<Vec<_>>()
     }
 
-    fn get_best_ip<'b>(
+    fn get_best_ip<'a>(
         &self,
-        ip_list: &'b [IpAddr],
+        ip_list: &'a [IpAddr],
         proxy_count_validated: bool,
         proxy_list_validated: bool,
-    ) -> (Option<&'b IpAddr>, bool) {
+    ) -> (Option<&'a IpAddr>, bool) {
         if ip_list.is_empty() {
             return (None, false);
         }
@@ -752,7 +754,7 @@ mod tests_ipv4_proxy_count {
     #[test]
     fn singleton_proxy_count() {
         let proxies = vec![];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
         let mut headers = HeaderMap::new();
         headers.insert("HTTP_X_FORWARDED_FOR", "177.139.233.139".parse().unwrap());
         let (ip_addr, trusted_route) = ipware.get_client_ip(&headers, false);
@@ -763,7 +765,7 @@ mod tests_ipv4_proxy_count {
     #[test]
     fn singleton_proxy_count_private() {
         let proxies = vec![];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
         let mut headers = HeaderMap::new();
         headers.insert("HTTP_X_FORWARDED_FOR", "10.0.0.0".parse().unwrap());
         headers.insert("X_REAL_IP", "177.139.233.139".parse().unwrap());
@@ -775,7 +777,7 @@ mod tests_ipv4_proxy_count {
     #[test]
     fn proxy_count_relax() {
         let proxies = vec![];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
         let mut headers = HeaderMap::new();
         headers.insert(
             "HTTP_X_FORWARDED_FOR",
@@ -791,7 +793,7 @@ mod tests_ipv4_proxy_count {
     #[test]
     fn proxy_count_strict() {
         let proxies = vec![];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
         let mut headers = HeaderMap::new();
         headers.insert(
             "HTTP_X_FORWARDED_FOR",
@@ -818,7 +820,7 @@ mod tests_ipv4_proxy_list {
             "198.84.193.157".parse::<IpAddr>().unwrap(),
             "198.84.193.158".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
 
         let mut headers = HeaderMap::new();
         headers.insert(
@@ -838,7 +840,7 @@ mod tests_ipv4_proxy_list {
             "198.84.193.157".parse::<IpAddr>().unwrap(),
             "198.84.193.158".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
 
         let mut headers = HeaderMap::new();
         headers.insert(
@@ -858,7 +860,7 @@ mod tests_ipv4_proxy_list {
             "198.84.193.157".parse::<IpAddr>().unwrap(),
             "198.84.193.158".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
 
         let mut headers = HeaderMap::new();
         headers.insert(
@@ -885,7 +887,7 @@ mod tests_ipv4_proxy_count_proxy_list {
             "198.84.193.157".parse::<IpAddr>().unwrap(),
             "198.84.193.158".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(2, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(2, proxies));
 
         let mut headers = HeaderMap::new();
         headers.insert(
@@ -905,7 +907,7 @@ mod tests_ipv4_proxy_count_proxy_list {
             "198.84.193.157".parse::<IpAddr>().unwrap(),
             "198.84.193.158".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(2, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(2, proxies));
 
         let mut headers = HeaderMap::new();
         headers.insert(
@@ -1183,7 +1185,7 @@ mod tests_ipv6_proxy_count {
     #[test]
     fn singleton_proxy_count() {
         let proxies = vec![];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
         let mut headers = HeaderMap::new();
         headers.insert(
             "HTTP_X_FORWARDED_FOR",
@@ -1198,7 +1200,7 @@ mod tests_ipv6_proxy_count {
     #[test]
     fn singleton_proxy_count_private() {
         let proxies = vec![];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(1, proxies));
         let mut headers = HeaderMap::new();
         headers.insert("HTTP_X_FORWARDED_FOR", "::1".parse().unwrap());
         headers.insert(
@@ -1225,7 +1227,7 @@ mod tests_ipv6_proxy_list {
             "2606:4700:4700::1111".parse::<IpAddr>().unwrap(),
             "2001:4860:4860::8888".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
         let mut headers = HeaderMap::new();
         headers.insert(
             "HTTP_X_FORWARDED_FOR",
@@ -1248,7 +1250,7 @@ mod tests_ipv6_proxy_list {
             "2606:4700:4700::1111".parse::<IpAddr>().unwrap(),
             "2001:4860:4860::8888".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
         let mut headers = HeaderMap::new();
         headers.insert(
             "HTTP_X_FORWARDED_FOR",
@@ -1271,7 +1273,7 @@ mod tests_ipv6_proxy_list {
             "2606:4700:4700::1111".parse::<IpAddr>().unwrap(),
             "2001:4860:4860::8888".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
         let mut headers = HeaderMap::new();
         headers.insert(
             "HTTP_X_FORWARDED_FOR",
@@ -1294,7 +1296,7 @@ mod tests_ipv6_proxy_list {
             "2606:4700:4700::1111".parse::<IpAddr>().unwrap(),
             "2001:4860:4860::8888".parse::<IpAddr>().unwrap(),
         ];
-        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, &proxies));
+        let ipware = IpWare::new(IpWareConfig::default(), IpWareProxy::new(0, proxies));
         let mut headers = HeaderMap::new();
         headers.insert(
             "HTTP_X_FORWARDED_FOR",

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -335,11 +335,12 @@ impl<'a> IpWare<'a> {
         }
     }
 
-    fn get_meta_values<'b>(&self, headers: &'b HeaderMap) -> Vec<&'b HeaderValue> {
+    fn get_meta_values(&self, headers: &'a HeaderMap) -> Vec<&'a str> {
         self.config
             .precedence
             .iter()
             .filter_map(|header_name| self.get_meta_value(headers, header_name))
+            .filter_map(|header_value| header_value.to_str().ok())
             .collect::<Vec<_>>()
     }
 
@@ -348,7 +349,7 @@ impl<'a> IpWare<'a> {
         let mut private_list = vec![];
         let meta_values = self.get_meta_values(headers);
         for &meta_value in meta_values.iter() {
-            let meta_ips = self.get_ips_from_string(meta_value.to_str().unwrap().to_owned());
+            let meta_ips = self.get_ips_from_string(meta_value);
             if meta_ips.is_empty() {
                 continue;
             }
@@ -390,7 +391,7 @@ impl<'a> IpWare<'a> {
     ///
     /// # Arguments
     /// * `ip_str` - String contains , seperated ip addresses
-    fn get_ips_from_string(&self, ip_str: String) -> Vec<IpAddr> {
+    fn get_ips_from_string(&self, ip_str: &str) -> Vec<IpAddr> {
         let mut result = ip_str
             .split(',')
             .map(|single_ip| {

